### DISCUSSION TITLE
[Trouble understanding semantics.ml][Hw2]

### DISCUSSION THREAD

**Discussion by justin-gonzaga** [prosyslab-classroom/cs524-program-analysis#422] [Q&A] [answered]
I am working on the Homework 2 assignment but I am having some trouble understanding what to do in the semantics.ml implementation. 
I was wondering if you could provide some hints on how to start?

I am mainly confused about how to use the modules in domain.ml and how to use the LLVM data structure and APIs. (I've also read the documentation but am still confused...https://github.com/prosyslab-classroom/llvm-primer/blob/master/data-structure.md https://llvm.moe/ocaml/Llvm.html). 

Based on this issue from last year https://github.com/prosyslab-classroom/cs524-program-analysis/issues/21, the types of each function have been provided:

eval : Llvm.llvalue -> Memory.t -> Value.t
execute_block : (Llvm.llbasicblock * Memory.t) option -> Llvm.llbasicblock -> Memory.t -> Memory.t
execute_instr : (Llvm.llbasicblock * Memory.t) option -> (Llvm.llbasicblock, Llvm.llvalue) Llvm.llpos -> Memory.t -> Memory.t
transfer : (Llvm.llbasicblock * Memory.t) option -> Llvm.llvalue -> Memory.t -> Memory.t

But I can't find any API functions that I can use such that my implementation matches the provided function type. For example, for the eval function, there doesn't seem to be any way to output a Value given an Llvm.llvalue. 

I am also not sure how to proceed with the implementation for the transfer function. 
For example, for the input instruction match case, my reasoning is like this:

1. Read in the user input. 
2. Update the memory with the input.
3. Execute the next instruction (by calling the execute_instr function)

But even after reading the documentation I am not sure how to implement these steps. 
For example, to update the memory, would you need to use the Memory.add() API function? 
And how would you call the execute_instr function from within the transfer function? Since the type of instr in the transfer function is different from the expected type in the execute_instr function.

I am also wondering if you could clarify the wording for the PHI implementation 
(* copy the value of the incoming variable from the incoming memory to the destination variable *)
So far my assumption is that the incoming memory is denoted by incoming_mem, the incoming variable is denoted by incoming_blk, and 
the destination variable is curr_blk? But if that's the case then I can't think of a way to use incoming_exp that is also in the implementation.

Thanks

**Comment by goodtaeeun** [comment, marked as answer]
Let me try to organize your question into 3 parts and answer them respectively.

1. Can't find any API functions
Let me answer this with the example you gave us. In short, there is no API function that directly outputs a `Value` type given an `Llvm.llvalue`. This is the very part that you need to implement by yourself. To give a hint, try to decompose the given `Llvm.llvalue` to compose a new Value type with it. Try using the API functions in https://llvm.moe/ocaml/Llvm.html.

2. How to implement the transfer function.
First of all, your reasoning for the implementation is correct. And yes, you would need to use `Memory.add` to update the memory. Finally, `instr` in `execute_instr` and `transfer` function isn't exactly the same thing. They simply have the same name. As you noted, their types are different so you must perform some kind of processing job to convert it.

3. PHI node
`curr_blk` is the basic block that holds the current instruction. However, the destination of the phi instruction is the current instruction itself, not the block. Note that phi instruction can have multiple values depending on the previous location, which is denoted by the incoming block. So you must find which variable is actually being used by referring to the incoming block.
For example, given a phi instruction as follows,
`%x = phi i32 [ %inc, %then ], [ %dec, %else ]`,
`%x` can be assigned with `%inc` if the previous location, or incoming block, was `%then`, or `%dec` if the previous location was `%else`.
Now you may guess what `incoming_exp` is.

Good luck.
Tae Eun.
